src/gdsync.cpp

```diff
@@ -1441,20 +1479,34 @@ struct gds_qp *gds_create_qp(struct ibv_pd *pd, struct ibv_context *context, gds
         qp_attr->comp_mask |= IBV_EXP_QP_INIT_ATTR_PEER_DIRECT;
         qp_attr->peer_direct_attrs = peer_attr;
 
+        gds_dbg("Comp Mask: %x, exp_create_flags: %x\n", qp_attr->comp_mask, qp_attr->exp_create_flags);
         qp = ibv_exp_create_qp(context, qp_attr);
         if (!qp)  {
                 ret = EINVAL;
-                gds_err("error in ibv_exp_create_qp\n");
+                gds_err("error in ibv_exp_create_qp, errno: %d, %s\n", errno, strerror(errno));
                 goto err_free_cqs;
         }
 
+        if(!is_qp_flusher)
+                gds_warn("created gds_qp=%p\n", gqp);
+        else
+                gds_warn("created flusher gds_qp=%p\n", gqp);
+
         gqp->qp = qp;
         gqp->send_cq.cq = qp->send_cq;
         gqp->send_cq.curr_offset = 0;
         gqp->recv_cq.cq = qp->recv_cq;
         gqp->recv_cq.curr_offset = 0;
 
-        gds_dbg("created gds_qp=%p\n", gqp);
+        if(!is_qp_flusher)
+        {
+                if(gds_flusher_init(pd, context, gpu_id))
```

Comment:
gds_flusher_init() should return a flusher object which is stored in gds_qp.
you should convince the reviewer that there is value in abstracting the native flusher inside , or to simply special case in gdsync.c